File: chat-backend/src/main/java/com/smallchat/backend/chat/application/usecase/ParticipatingChatsUseCase.java
package com.smallchat.backend.chat.application.usecase;

import java.util.List;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.smallchat.backend.chat.application.inputport.LastChatMessageInputPort;
import com.smallchat.backend.chat.application.inputport.ParticipatingChatsInputPort;
import com.smallchat.backend.chat.application.outputport.ChatOutputPort;
import com.smallchat.backend.chat.domain.model.Chat;
import com.smallchat.backend.chat.domain.model.vo.Message;
import com.smallchat.backend.chat.framework.web.dto.ChatBasicInfoListDto;
import com.smallchat.backend.chat.framework.web.dto.ChatDetail;
import com.smallchat.backend.user.application.inputport.UserChatListInputPort;
import com.smallchat.backend.user.domain.model.ParticipatingChat;

import lombok.RequiredArgsConstructor;

@Service
@RequiredArgsConstructor
public class ParticipatingChatsUseCase implements ParticipatingChatsInputPort {

    //todo inputport 수정 필요
    private final UserChatListInputPort userChatListInputPort;
    private final LastChatMessageInputPort lastChatMessageInputPort;
    private final ChatOutputPort chatOutputPort;

    @Override
    @Transactional
    public ChatBasicInfoListDto.Response getChatList(String userId) {
        List<ParticipatingChat> userJoinedChats = userChatListInputPort.getUserJoinedChats(userId);
        List<String> userJoinedChatIdList = userJoinedChats.stream().map(ParticipatingChat::getChatId).toList();

        List<Chat> chatList = chatOutputPort.findChatBasicByIds(userJoinedChatIdList);
        List<Message> lastMessageList = lastChatMessageInputPort.getLastMessageList(userJoinedChatIdList);

        return new ChatBasicInfoListDto.Response(chatList.stream().map(el -> el.toChatBasicInfo(lastMessageList)).toList());
    }

    @Override
    public ChatDetail.Response getChatDetail(String chatId) {
        Chat chat = chatOutputPort.load(chatId);
        return new ChatDetail.Response(chatId, chat.getName());
    }
}

File: chat-backend/src/main/java/com/smallchat/backend/chat/application/usecase/ParticipatingChatsUseCase.kt
package com.smallchat.backend.chat.application.usecase

import com.smallchat.backend.chat.domain.interfaces.ChatRepository
import com.smallchat.backend.chat.domain.interfaces.ChatUserRepository
import com.smallchat.backend.chat.domain.interfaces.MessageRepositoryKt
import com.smallchat.backend.chat.interfaces.web.dto.ChatDetail
import com.smallchat.backend.chat.interfaces.web.dto.ChatListDto
import com.smallchat.backend.chat.interfaces.web.dto.ChatMessage
import jakarta.persistence.EntityNotFoundException
import org.springframework.stereotype.Service

@Service
class ParticipatingChatsUseCase(
    val chatUserRepository: ChatUserRepository,
    val messageRepositoryKt: MessageRepositoryKt,
    val chatRepository: ChatRepository
) {

    fun getChatList(userId: String): ChatListDto.Response {
        val chatUsers = chatUserRepository.findAllByUserId(userId)
        val chatMessages = chatUsers.map { chatUser ->
            val chatId = chatUser.chat.chatIdOrThrow
            val chatName = chatUser.chat.name
            val lastMessageKt =
                messageRepositoryKt.findFirstByChatIdOrderBySentAtDesc(chatId) ?: throw IllegalStateException("why")
            ChatMessage(chatId, chatName, lastMessageKt.message, lastMessageKt.actualSentAt)
        }

        return ChatListDto.Response(chatMessages)
    }

    fun getChatDetail(chatId: String): ChatDetail.Response {
        val chat = chatRepository.findById(chatId).orElseThrow { throw EntityNotFoundException("찾을 수 없는 채팅 엔티티") }
        return ChatDetail.Response(chatId, chat.name)
    }
}
